packages/components/src/utils/hooks/use-focus-outside.js

```diff
@@ -0,0 +1,211 @@
+ * element the event handlers are bound to.
+ *
+ * @param {EventCallback} onFocusOutside        A callback triggered when focus moves outside
+ *                                              the element the event handlers are bound to.
+ * @param {FocusOutsideRef} [__unstableNodeRef]
+ *
+ * @return {FocusOutsideReturnValue} An object containing event handlers. Bind the event handlers
+ *                                   to a wrapping element element to capture when focus moves
+ *                                   outside that element.
+ */
+export default function useFocusOutside( onFocusOutside, __unstableNodeRef ) {
+	const preventBlurCheck = useRef( false );
+
+	/**
+	 * @type {import('react').MutableRefObject<number | undefined>}
+	 */
+	const blurCheckTimeoutId = useRef();
+
+	/**
+	 * Cancel a blur check timeout.
+	 */
+	const cancelBlurCheck = useCallback( () => {
+		clearTimeout( blurCheckTimeoutId.current );
+	}, [ blurCheckTimeoutId ] );
+
+	// Cancel blur checks on unmount.
+	useEffect( () => {
+		return () => cancelBlurCheck();
+	}, [] );
+
+	// Cancel a blur check if the callback or ref is no longer provided.
+	useEffect( () => {
+		if ( ! onFocusOutside && ! __unstableNodeRef?.current ) {
+			cancelBlurCheck();
+		}
+	}, [ onFocusOutside, __unstableNodeRef ] );
+
+	/**
+	 * Handles a mousedown or mouseup event to respectively assign and
+	 * unassign a flag for preventing blur check on button elements. Some
+	 * browsers, namely Firefox and Safari, do not emit a focus event on
+	 * button elements when clicked, while others do. The logic here
+	 * intends to normalize this as treating click on buttons as focus.
+	 *
+	 * @see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#Clicking_and_focus
+	 *
+	 * @param {SyntheticEvent} event Event for mousedown or mouseup.
+	 */
+	const normalizeButtonFocus = useCallback(
+		( event ) => {
+			const { type, target } = event;
+			const isInteractionEnd = includes(
+				[ 'mouseup', 'touchend' ],
+				type
+			);
+
+			if ( isInteractionEnd ) {
+				preventBlurCheck.current = false;
+			} else if ( isFocusNormalizedButton( target ) ) {
+				preventBlurCheck.current = true;
+			}
+		},
+		[ preventBlurCheck ]
+	);
+
+	/**
+	 * A callback triggered when a blur event occurs on the element the handler
+	 * is bound to.
+	 *
+	 * Calls the `onFocusOutside` callback in an immediate timeout if focus has
+	 * move outside the bound element and is still within the document.
+	 *
+	 * @param {SyntheticEvent} event Blur event.
+	 */
+	const queueBlurCheck = useCallback(
+		( event ) => {
+			// React does not allow using an event reference asynchronously
+			// due to recycling behavior, except when explicitly persisted.
+			event.persist();
+
+			// Skip blur check if clicking button. See `normalizeButtonFocus`.
+			if ( preventBlurCheck.current ) {
+				return;
+			}
+
+			blurCheckTimeoutId.current = setTimeout( () => {
```

Comment:
I wonder if we can change `setTimeout` here to something more semantic? Like maybe `rAF`? I understand that if we want to preserve backward compatibility first though.

Reply:
Not sure, I feel like the saying "if it ain't broke, don't fix it" applies here. There'd be little discernable value from switching to `rAF`, when the timeout has been working successfully for a couple of years.

It's quite hard to predict what consequences changing it might have as well, so if it were to be changed it should be in a separate PR. That way, if there's a bug caused by the change it'll be easy to bisect, easy to reason about and easy to revert.